DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,273 @@
+          unsigned k_raw_latency,   // RAW latency for triangular loop
+                                    // optimization in the QR iteration
+          int k_zero_threshold_1e,  // Threshold from which we consider a
+          typename T                // The datatype for the computation
+          >
+double singularValueDecomposition(
+    std::vector<T> &a_matrix,  // input matrix A (col maj)
+    std::vector<T> &u_matrix,  // left singular vectors, U (col maj)
+    std::vector<T> &s_matrix,  // list of singular values, S
+    std::vector<T> &v_matrix,  // right singular vectors, V (row maj)
+    std::vector<ac_int<1, false>>
+        &rank_deficient_flag,  // Output a flag that is 1 if the input matrix
+                               // is considered rank deficient
+    sycl::queue &q, int repetitions) {
+  constexpr int kNumElementsPerDDRBurst = 8;            // total of 256 bit
+  constexpr int kInputMatrixSize = rows * cols;         // row x col
+  constexpr int kEigenVectorsMatrixSize = cols * cols;  // col x col
+  constexpr int kUMatrixSize = rows * rows;
+  constexpr int kRMatrixSize = rows * (rows + 1) / 2;
+
+  // pipes
+  using PipeType = fpga_tools::NTuple<T, kNumElementsPerDDRBurst>;
+
+  using InputMatrixPipe = sycl::ext::intel::pipe<IMP, PipeType, 16>;
+  using InputMatrixPipe2 = sycl::ext::intel::pipe<IMP2, PipeType, 16>;
+  using CovarianceMatrixPipe = sycl::ext::intel::pipe<CMP, PipeType, 16>;
+  using EigenValuesPipe = sycl::ext::intel::pipe<EValP, T, 3>;
+  using EigenVectorsPipe = sycl::ext::intel::pipe<EVecP, PipeType, 3>;
+  using RankDeficientFlagPipe =
+      sycl::ext::intel::pipe<RDFP, ac_int<1, false>, 3>;
+  using UTempMatrixPipe = sycl::ext::intel::pipe<OUTMP, PipeType, 3>;
+  using UMatrixPipe = sycl::ext::intel::pipe<OUMP, PipeType, 3>;
+  using RMatrixPipe =
+      sycl::ext::intel::pipe<ORMP, T, kNumElementsPerDDRBurst * 4>;
+  using SMatrixPipe = sycl::ext::intel::pipe<OSMP, PipeType, 3>;
+  using VMatrixPipe = sycl::ext::intel::pipe<OVMP, PipeType, 3>;
+
+  // USM allocations
+  T *input_matrix_device;
+  ac_int<1, false> *rank_deficient_flag_device;
+  T *u_matrix_device;
+  T *s_matrix_device;
+  T *v_matrix_device;
+  T *r_device;
+
+  if (q.get_device().has(sycl::aspect::usm_device_allocations)) {
+    std::cout << "Using device allocations" << std::endl;
+    // Allocate FPGA DDR memory.
+    input_matrix_device = sycl::malloc_device<T>(kInputMatrixSize, q);
+    rank_deficient_flag_device = sycl::malloc_device<ac_int<1, false>>(1, q);
+    u_matrix_device = sycl::malloc_device<T>(kUMatrixSize, q);
+    s_matrix_device = sycl::malloc_device<T>(kInputMatrixSize, q);
+    v_matrix_device = sycl::malloc_device<T>(kEigenVectorsMatrixSize, q);
+    r_device = sycl::malloc_device<T>(kRMatrixSize, q);
+  } else if (q.get_device().has(sycl::aspect::usm_shared_allocations)) {
+    std::cout << "Using shared allocations" << std::endl;
+    // No device allocations means that we are probably in a SYCL HLS flow
```

Comment:
```suggestion
    // No device allocations means that we are probably in a SYCL HLS flow
```
I disagree with this comment; no device allocations only means that the BSP does not support device allocations. Also, I don't see how the code below specifically leverages SYCL HLS in any way, so there isn't any point in mentioning this.